compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp

```diff
@@ -463,16 +562,29 @@ static Value getCombiningIdentityValue(Location loc, OpBuilder &builder,
   return identity;
 }
 
+/// Find function op and check it has redux support
+static bool hasReduxSupport(Operation *op) {
+  while (op) {
+    if (auto funcOp = dyn_cast<func::FuncOp>(op)) {
+      return getTargetInfo(funcOp).hasRedux;
+    }
+    op = op->getParentOp();
+  }
+  return false;
+}
+
 /// Emit reduction across a group for a given input.
 Value emitGPUGroupReduction(Location loc, OpBuilder &builder, Value input,
                             vector::CombiningKind kind, uint32_t size,
                             const int warpSize) {
   assert(
       size % warpSize == 0 &&
       "Group reduction only support for sizes aligned on warp size for now.");
+  bool hasRedux = hasReduxSupport(input.getDefiningOp());
```

Comment:
Yea we can. In this case, I always generate a subgroupReduceOp here. I move codes in 'warpReduction' in a lowering pattern subgroupReduce op. Is that what we what? 

Bte where should subgroupReduceOp lowering live? In MLIR core?

Reply:
> Yea we can. In this case, I always generate a subgroupReduceOp here. I move codes in 'warpReduction' in a 
lowering pattern subgroupReduce op. Is that what we what?

Yes 

> Bte where should subgroupReduceOp lowering live? In MLIR core?

Yes in case it doesn't exist yet. Could be a gpu -> gpu pattern or a gpu -> nvvm one (the gpu to gpu might be more interesting)

Reply:
I postponed lowering of `gpu.subgroup_reduce` op. Here what happens now

**default case for nvvm:**
 reduction code -> `gpu.subgroup_reduce` -> `gpu.shuffle + arith` -> `nvvm.shuffle + arith`

**redux enabled for nvvm:**
 reduction code -> `gpu.subgroup_reduce` -> `gpu.subgroup_reduce` -> `nvvm.redux`